### PR TITLE
[refactor] Switch to using Jargo instead of the defunct Apache Excalibur

### PR DESCRIPTION
**Pull request by adamretter**
Modernises command line argument parsing

Closes https://github.com/eXist-db/exist/issues/1195

### CODE REVIEW

**Comment by dizzzz** [comment]
Does the jargon lib require then google library? How large is it?  I typically use another library, can't recall the name at his moment. Smaller and not a 0.x version :/

**Comment by adamretter** [comment]
@dizzzz It was a temporary problem with Travis being unable to access the web to retrieve the betterForm dependencies. I just restarted the build.

Yes, Jargo depends on Google Guava. I looked at a few different options but felt that this one was the best fit. Having done this, I rather like its syntax.

**Comment by dizzzz** [comment]
We need to update the list of jarfiles used by the webstart client too....

Actually the current list is too long, maybe @wolfgangmm can have a look?

**Comment by dizzzz** [comment]
I often use Jcommander .... no dependencies , very mature , nice syntax

**Comment by adamretter** [comment]
@dizzzz Jcommander has a few design issues which Jargo improves on, see: https://github.com/Softhouse/jargo#rationale

**Comment by shabanovd** [comment]
welcome to guava dependency hell -) ... latest version is 20.0, no?

**Comment by adamretter** [comment]
@shabanovd Yes the latest is 20.0, but right now I only need it for Jargo which is built against 14.0.1. We could update in future if we need something else, but right now we don't.

**Comment by shabanovd** [comment]
why introduce dependency on 3 years version? If it's because of possible issue with Jargo why use it?

**Comment by adamretter** [comment]
@shabanovd Do you have a concrete problem for me to address?

**Comment by shabanovd** [comment]
Yes, update guava to latest version.

**Comment by dizzzz** [comment]
Without diving into details, I know a lot of ppl use guava, but in these days a lot is in java8 no? I have never used it because I gout the impression it is powerful but overly complex .....

Re comparison jcommander... sure there are some imperfections but at least there is no dependency with guava....

**Comment by adamretter** [comment]
@shabanovd As we are not using Guava for anything else, I do not see the point in undertaking the work to update the version.

@dizzzz Having done this work with this library which I selected because I believed it to offer the best functionality, I am not wanting to redo that work.

Guys! This PR fixes a concrete issue that a user reported, can we please just move forward rather than worrying about what *might* be rather than *what is*. If people want to use Guava in future for other things, they are welcome to send a PR to update it to a newer version (if they need a newer version)

**Comment by dizzzz** [comment]
Ok, pull it in then :)

**Comment by dizzzz** [comment]
Ok, pull it in, ok @wolfgangmm ?

**Comment by dizzzz** [comment]
If @wolfgangmm agrees?

**Comment by shabanovd** [comment]
@adamretter eXist is used as embedded database. Some projects may use Guava. Here the [link](https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#newwindow=1&q=guava+version+conflicts) that give you understand how that Guava world looks like.

**Comment by adamretter** [comment]
@shabanovd This still doesn't seem concrete:

> Some projects may use Guava

We could say the same of any library that we add to eXist. Do you have any concrete example of a project which embeds eXist and uses a newer incompatible version of Guava?

**Comment by shabanovd** [comment]
as example eD use 19.0-rc1

**Comment by joewiz** [comment]
eD? Can you provide a link?

**Comment by shabanovd** [comment]
sure, [here it is](http://easydita.com/)

**Comment by joewiz** [comment]
Ah, I hadn't heard of that acronym for easyDITA, but it makes sense. Thanks!

**Comment by adamretter** [comment]
@shabanovd Okay then, I will look at updating it upstream: https://github.com/Softhouse/jargo/pull/34

**Comment by jontejj** [comment]
Hey guys! I have not maintained the jargo lib for a while but it would be interesting to switch it to use Java 8 features instead of Guava. Updating the version is of course an easier fix but give me a couple of weeks and I might pull it off :)

**Comment by adamretter** [comment]
@jontejj I think you might struggle to find replacements for some of the Guava stuff you are using in Java 8, but I will be interested to see your conclusions :-)

**Comment by adamretter** [comment]
@jontejj any news? Or should we ship this PR?

**Comment by jontejj** [comment]
No news I'm afraid. Been totally overloaded with work since christmas. Will update the Guava version now. Sorry for dropping the ball :)

**Comment by adamretter** [comment]
@jontejj Great thanks

**Comment by adamretter** [comment]
@shabanovd As requested this now uses Guava 20.0 as requested. Thanks to @jontejj for updating Jargo for us :-)

@dizzzz This is now ready to be merged

**Comment by dizzzz** [comment]
At least the webstart list must be updated, I realize. The list is now too long, need help te review what is needed for a remote collection only.....

**Comment by adamretter** [comment]
I don't really understand how webstart works. Where is the list?

**Comment by shabanovd** [comment]
@adamretter [here](https://github.com/eXist-db/exist/blob/9a380a6730e9d77e3d7bda0bae3b4b22d2b727cb/src/org/exist/webstart/JnlpJarFiles.java#L51)

**Comment by dizzzz** [comment]
Pretty sure explained you previously :)

**Comment by adamretter** [comment]
@shabanovd @dizzzz Thanks, I made the changes to WebStart for this PR as requested.

**Comment by jontejj** [comment]
@adamretter I made a branch of jargo without Guava (https://github.com/jontejj/jargo/pull/8). I would love your feedback on it!

**Comment by dizzzz** [comment]
👍🏼👍🏼

**Comment by adamretter** [comment]
@jontejj Yeah that seems to work for me :-) Thank you very very much.
Do you plan to publish a release?

**Comment by jontejj** [comment]
Yeah, once I have polished the documentation a bit.

**Comment by adamretter** [comment]
@jontejj Amazing, thank you again.

**Comment by adamretter** [comment]
@dizzzz @shabanovd This is ready to land. It is now using a snapshot of the upcoming Jargo-0.4.2 which is Guava free as you requested.

File: src/org/exist/backup/Main.java [hunk 1]
```diff
@@ -65,159 +62,120 @@
...
+
+            } finally {
+                if (pin != null) {
                     pin.close();
                 }
             }
-
-        } catch (final IOException ioe) {
+        } catch (final IOException e) {
+            e.printStackTrace();
```
Comment (shabanovd) [inline src/org/exist/backup/Main.java]: `System.exit`?

File: src/org/exist/backup/Main.java [hunk 2]
```diff
@@ -459,9 +402,20 @@ private static void shutdown(Collection root) {
...
+                    .withArguments(userArg, passwordArg, dbaPasswordArg)
+                    .andArguments(backupCollectionArg, backupOutputDirArg)
+                    .andArguments(restoreArg, rebuildExpathRepoArg)
+                    .andArguments(helpArg, guiArg, quietArg, optionArg)
+                    .parse(args);
+
+            process(arguments);
+        } catch(final ArgumentException e) {
+            System.out.println(e.getMessageAndUsage());
+            System.exit(2);
```
Comment (shabanovd) [inline src/org/exist/backup/Main.java]: exit code `128`?

File: src/org/exist/jetty/ServerShutdown.java [hunk 3]
```diff
@@ -20,83 +14,81 @@
...
-            new CLOptionDescriptor("password",
-                    CLOptionDescriptor.ARGUMENT_REQUIRED, PASSWORD_OPT,
-                    "specify password for the user."),
-            new CLOptionDescriptor("uri", CLOptionDescriptor.ARGUMENT_REQUIRED,
-                    URI_OPT,
-                    "the XML:DB URI of the database instance to be shut down.") };
+            process(arguments);
+        } catch (final ArgumentException e) {
+            System.out.println(e.getMessageAndUsage());
+            System.exit(2);
```
Comment (shabanovd) [inline src/org/exist/jetty/ServerShutdown.java]: `128`?

File: src/org/exist/jetty/ServerShutdown.java [hunk 4]
```diff
@@ -120,35 +112,30 @@ public static void main(String[] args) {
...
-            	try {
-            		clientProps.load(pin);
-            	} finally {
-            		pin.close();
-            	}
-            } catch (final IOException ex) {
-                //
+                final Path propFile = ConfigurationHelper.lookup("client.properties");
+
+                if (Files.isReadable(propFile)) {
```
Comment (shabanovd) [inline src/org/exist/jetty/ServerShutdown.java]: method + `try-with-resources`

File: src/org/exist/jetty/ServerShutdown.java [hunk 5]
```diff
@@ -120,35 +112,30 @@ public static void main(String[] args) {
...
+                    properties.load(pin);
+                }
+
+            } finally {
+                if (pin != null) {
+                    pin.close();
+                }
             }
+        } catch (final IOException e) {
+            e.printStackTrace();
```
Comment (shabanovd) [inline src/org/exist/jetty/ServerShutdown.java]: `System.exit`?

File: test/src/org/exist/performance/Main.java [hunk 6]
```diff
@@ -34,77 +30,70 @@
...
+            final ParsedArguments arguments = CommandLineParser
+                    .withArguments(reportArg, inputFileArg, outputDirArg)
+                    .andArguments(groupArg)
+                    .andArguments(helpArg)
+                    .parse(args);
+
+            process(arguments);
+        } catch (final ArgumentException e) {
+            System.out.println(e.getMessageAndUsage());
+            System.exit(2);
```
Comment (shabanovd) [inline test/src/org/exist/performance/Main.java]: `128`?

File: test/src/xquery/TestRunnerMain.java [hunk 7]
```diff
@@ -47,67 +49,65 @@
...
-            return;
-        }
-        List<String> files = new ArrayList<String>();
-        String id = null;
-        
-        List<CLOption> opts = optParser.getArguments();
+			process(arguments);
+		} catch (final ArgumentException e) {
+			System.out.println(e.getMessageAndUsage());
+			System.exit(2);
```
Comment (shabanovd) [inline test/src/xquery/TestRunnerMain.java]: `128`?

File: src/org/exist/util/ArgumentUtil.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+     *
+     * @param parsedArguments The arguments which have been parsed
+     * @param argument The option argument that we are looking for
+     *
+     * @return true if the option was set, false otherwise
+     */
+    public static boolean getBool(final ParsedArguments parsedArguments, final Argument<Boolean> argument) {
+        return getOpt(parsedArguments, argument)
+                .flatMap(Optional::ofNullable)
+                .orElse(false);
```
Comment (shabanovd) [inline src/org/exist/util/ArgumentUtil.java]: better to define `default` parameter to make it clear

File: src/org/exist/management/client/JMXClient.java [hunk 9]
```diff
@@ -247,129 +244,114 @@ private void error(Exception e) {
...
+                    .withArguments(addressArg, portArg, instanceArg, waitArg)
+                    .andArguments(cacheDisplayArg, locksDisplayArg)
+                    .andArguments(dbInfoArg, memoryInfoArg, sanityCheckInfoArg, jobsInfoArg)
+                    .andArguments(helpArg)
+                    .parse(args);
+
+            process(arguments);
+        } catch (final ArgumentException e) {
+            System.out.println(e.getMessageAndUsage());
+            System.exit(2);
```
Comment (shabanovd) [inline src/org/exist/management/client/JMXClient.java]: `128`?

File: src/org/exist/client/InteractiveClient.java [hunk 10]
```diff
@@ -246,10 +238,18 @@ protected void displayHelp() {
...
+            final CommandlineOptions options = CommandlineOptions.parse(args);
+
+            final InteractiveClient client = new InteractiveClient(options);
+            if (!client.run()) {
                 System.exit(1); // return non-zero exit status on failure
             }
+
+        } catch(final ArgumentException e) {
+            System.out.println(e.getMessageAndUsage());
+            System.exit(2);
```
Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: `128`?

File: src/org/exist/client/InteractiveClient.java [hunk 11]
```diff
@@ -1795,226 +1829,68 @@ protected Collection getCollection(final String path) throws XMLDBException {
...
-        }
+            InputStream pin = null;
+            try {
+                final Path propFile = ConfigurationHelper.lookup("client.properties");
 
-        if (pin == null) {
-            // Try to load via classloader
-            pin = InteractiveClient.class.getResourceAsStream("client.properties");
-        }
+                if (Files.isReadable(propFile)) {
```
Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: method + `try-with-resources`

File: src/org/exist/backup/Main.java (hunk 1 above)
Comment (adamretter) [inline src/org/exist/backup/Main.java]: No, I wanted the fact the `backup.properties` file is missing to be a non-fatal error.

File: src/org/exist/backup/Main.java (hunk 2 above)
Comment (adamretter) [inline src/org/exist/backup/Main.java]: `128` is reserved for when you give an invalid number (i.e. `0 > n > 255`) to `exit`

File: src/org/exist/jetty/ServerShutdown.java (hunk 3 above)
Comment (adamretter) [inline src/org/exist/jetty/ServerShutdown.java]: `128` is reserved for when you give an invalid number (i.e. `0 > n > 255`) to `exit`

File: src/org/exist/jetty/ServerShutdown.java (hunk 5 above)
Comment (adamretter) [inline src/org/exist/jetty/ServerShutdown.java]: No, I wanted the fact the `client.properties` file is missing to be a non-fatal error.

File: test/src/org/exist/performance/Main.java (hunk 6 above)
Comment (adamretter) [inline test/src/org/exist/performance/Main.java]: 128 is reserved for when you give an invalid number (i.e. 0 > n > 255) to exit

File: src/org/exist/backup/Main.java (hunk 1 above)
Comment (shabanovd) [inline src/org/exist/backup/Main.java]: why printout trace then? just say that property file was not found.

File: test/src/xquery/TestRunnerMain.java (hunk 7 above)
Comment (adamretter) [inline test/src/xquery/TestRunnerMain.java]: 128 is reserved for when you give an invalid number (i.e. 0 > n > 255) to exit

File: src/org/exist/jetty/ServerShutdown.java (hunk 5 above)
Comment (shabanovd) [inline src/org/exist/jetty/ServerShutdown.java]: just say that `client.properties` file was not found, no needs for trace

File: src/org/exist/management/client/JMXClient.java (hunk 9 above)
Comment (adamretter) [inline src/org/exist/management/client/JMXClient.java]: 128 is reserved for when you give an invalid number (i.e. 0 > n > 255) to exit

File: src/org/exist/client/InteractiveClient.java (hunk 10 above)
Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: 128 is reserved for when you give an invalid number (i.e. 0 > n > 255) to exit

**Comment by adamretter** [comment]
@shabanovd Done. I have centralised and standardised the error codes used for `System.exist` in eXist. I have also switched those bits that you asked for to `try-with-resources`

**Comment by shabanovd** [comment]
@adamretter did you rebase? Comments are not attached to code anymore -(

File: test/src/xquery/TestRunnerMain.java [hunk 12]
```diff
@@ -47,67 +50,65 @@
...
 		try {
 			StringBuilder results = new StringBuilder();
 			XQueryService xqs = (XQueryService) rootCollection.getService("XQueryService", "1.0");
 			Source query = new FileSource(Paths.get("test/src/xquery/runTests.xql"), false);
-			for (String fileName : files) {
-				final Path file = Paths.get(fileName);
+			for (final Path file : files) {
 				if (!Files.isReadable(file)) {
-					System.console().printf("Test file not found: %s\n", fileName);
+					System.console().printf("Test file not found: %s\n", FileUtils.fileName(file));
```
Comment (shabanovd) [inline test/src/xquery/TestRunnerMain.java]: provide path as it

File: src/org/exist/jetty/ServerShutdown.java [hunk 13]
```diff
@@ -120,35 +113,22 @@ public static void main(String[] args) {
...
+                try(final InputStream pin = Files.newInputStream(propFile)) {
+                    properties.load(pin);
+                }
+            } else {
+                try(final InputStream pin = ServerShutdown.class.getResourceAsStream("client.properties")) {
+                    properties.load(pin);
+                }
             }
+        } catch (final IOException e) {
+            e.printStackTrace();
```
Comment (shabanovd) [inline src/org/exist/jetty/ServerShutdown.java]: stack trace will confuse end-user

**Comment by adamretter** [comment]
Yes I rebased, I always do so that I have nice clear PRs

File: src/org/exist/client/InteractiveClient.java [hunk 14]
```diff
@@ -1188,19 +1225,19 @@ private void remove(final String pattern) throws XMLDBException {
         }
     }
 
-    private void xupdate(final String resource, final String filename) throws XMLDBException, IOException {
-        final Path file = Paths.get(filename);
+    private void xupdate(final Optional<String> resource, final Path file) throws XMLDBException, IOException {
         if (!(Files.exists(file) && Files.isReadable(file))) {
-            messageln("cannot read file " + filename);
+            messageln("cannot read file " + FileUtils.fileName(file));
```
Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: provide full path

File: src/org/exist/client/InteractiveClient.java [hunk 15]
```diff
@@ -1341,7 +1375,7 @@ protected synchronized boolean parse(String fileName) throws XMLDBException {
                     files.add(file);
                 }
             } else {
-                files = DirectoryScanner.scanDir(fileName);
+                files = DirectoryScanner.scanDir(FileUtils.fileName(file));
```
Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: are you sure?

File: src/org/exist/client/InteractiveClient.java [hunk 16]
```diff
@@ -1795,226 +1829,60 @@ protected Collection getCollection(final String path) throws XMLDBException {
...
+                }
+            } else {
+                try(final InputStream pin = InteractiveClient.class.getResourceAsStream("client.properties")) {
+                    properties.load(pin);
                 }
-            } catch (final IOException ex) {
-                //
             }
+        } catch (final IOException e) {
+            e.printStackTrace();
```
Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: trace will confuse end-user

File: src/org/exist/client/CommandlineOptions.java [hunk 17]
```diff
@@ -21,157 +21,296 @@
...
+import se.softhouse.jargo.*;
+
+import static org.exist.util.ArgumentUtil.*;
+import static se.softhouse.jargo.Arguments.*;
 
 /**
+ * Command Line Options for the {@link InteractiveClient}
  *
- * @author wessels
+ * @author Adam Retter <adam.retter@googlemail.com>
```
Comment (shabanovd) [inline src/org/exist/client/CommandlineOptions.java]: @adamretter why did you remove `@author wessels`?

**Comment by shabanovd** [comment]
@adamretter "rebased" make reviewer live much harder especially with big changes PR like this one. Please, don't do that and address comments by separate commits.

File: test/src/xquery/TestRunnerMain.java (hunk 12 above)
Comment (adamretter) [inline test/src/xquery/TestRunnerMain.java]: Done

File: src/org/exist/jetty/ServerShutdown.java (hunk 13 above)
Comment (adamretter) [inline src/org/exist/jetty/ServerShutdown.java]: Okay I have made this more friendly

File: src/org/exist/client/InteractiveClient.java (hunk 14 above)
Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: Done

File: src/org/exist/client/InteractiveClient.java (hunk 15 above)
Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: Fixed

File: src/org/exist/client/InteractiveClient.java (hunk 16 above)
Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: Okay I have made this more friendly

File: src/org/exist/client/CommandlineOptions.java (hunk 17 above)
Comment (adamretter) [inline src/org/exist/client/CommandlineOptions.java]: Because this was a complete re-write and Dannes was not involved

**Comment by joewiz** [comment]
Perhaps if there is a rebase the submitter should post updated links to the relevant portion of code where comments were left? Then the reviewer can verify without additional labor.

**Comment by adamretter** [comment]
@joewiz @shabanovd GitHub does show you the comments that have been addressed with their "Show Outdated" links, no?

**Comment by adamretter** [comment]
@shabanovd Okay, changes are done.

**Comment by adamretter** [comment]
@joewiz Okay let's schedule such a discussion for the teleconf. Perhaps there should be a final rebase phase after review is approved, but before a merge?

File: src/org/exist/client/CommandlineOptions.java (hunk 17 above)
Comment (shabanovd) [inline src/org/exist/client/CommandlineOptions.java]: @adamretter did you copy anything from original file: text or code?

Comment (adamretter) [inline src/org/exist/client/CommandlineOptions.java]: I did not believe that any significant code was copied. But I am beyond caring and I just want this merged, and so I have added Dannes back in to make you happy.

Comment (shabanovd) [inline src/org/exist/client/CommandlineOptions.java]: @adamretter That is not me but copyright law. The work of others must be respected even if it's just a few words ;-)

Comment (adamretter) [inline src/org/exist/client/CommandlineOptions.java]: @shabanovd I do respect copyright law. As I said above, I believed this to be a rewrite.

**Comment by joewiz** [comment]
Sorry, yes, "show outdated" preserves the original comments; I was on the wrong track in my earlier comment. 

Rather, I think @shabanovd's point was that rebasing makes it hard to see the changes made *in response to a comment*, as separate from *all of the original work* in the PR. Having the newest changes isolated in separate commits helps the reviewer see what changed - and thus facilitates the work of the reviewer. 

I think your idea sounds promising, Adam. How about this: 

if the PR submitter (or the reviewer, for that matter) thinks there should be a final rebase (or, e.g., that the final PR should just be squashed and merged), this should be mentioned in the comment thread as commits are added in response to comments. For example, "I've addressed your questions in this commit, but before you merge it, I'd like to rebase. Let me know when this is ok."

Then it would be clear the PR shouldn't be merged before a final rebase is completed. 

I've added this to the [agenda for the Monday teleconf](https://docs.google.com/document/d/116O2kVM5_7QNsL7DVe21y-GFM7OQKi1FjfD8NLnqFXQ/edit?usp=sharing), under Agenda > 3c. Hopefully there will be time. 

It would be great to iron out everyone's assumptions and expectations on submitting and reviewing PRs.

**Comment by dizzzz** [comment]
I do not mind if my name disappeared here :)

**Comment by adamretter** [comment]
@dizzzz @shabanovd Okay so I have addressed all the concerns. Can we get it merged now please?

File: src/org/exist/client/InteractiveClient.java [hunk 18]
```diff
@@ -1309,18 +1346,15 @@ private synchronized boolean findRecursive(final Collection collection, final Pa
     /**
      * stores given Resource
      *
-     * @param fileName simple file or directory
+     * @param file file or directory
      * @throws XMLDBException
      */
-    protected synchronized boolean parse(String fileName) throws XMLDBException {
+    protected synchronized boolean parse(final Path file) throws XMLDBException {
```
Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: @adamretter you change behavior here:
was
```
fileName = args[1];
fileName = fileName.replace('/', java.io.File.separatorChar).replace('\\', java.io.File.separatorChar);		
final Path file = Paths.get(fileName).normalize();
```
become
```
Paths.get(args[1])
```

File: src/org/exist/client/InteractiveClient.java [hunk 19]
```diff
@@ -2213,23 +2086,22 @@ private void connectToDatabase() {
...
     }
 
     /**
      * Main processing method for the InteractiveClient object
      *
-     * @param args arguments from main()
      * @return true on success, false on failure
      */
-    public boolean run(final String args[]) throws Exception {
+    public boolean run() throws Exception {
```
Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: API change!

**Comment by dizzzz** [comment]
I'll review tomorrow...

File: src/org/exist/backup/ExportMain.java [hunk 20]
```diff
@@ -191,13 +164,11 @@ public static void main(String[] args) {
             }
 
             if (export) {
-                final Path dir = Paths.get(exportTarget);
-
-                if (!Files.exists(dir)) {
-                    Files.createDirectories(dir);
+                if (!Files.exists(exportTarget)) {
+                    Files.createDirectories(exportTarget);
```
Comment (dizzzz) [inline src/org/exist/backup/ExportMain.java]: maybe switch with next line?

File: src/org/exist/launcher/Launcher.java [hunk 21]
```diff
@@ -713,7 +714,7 @@ protected void showMessageAndExit(String title, String message, boolean logs) {
         utilityPanel.setVisible(true);
 
         JOptionPane.showMessageDialog(splash, panel, title, JOptionPane.WARNING_MESSAGE);
-        //System.exit(1);
+        //System.exit(SystemExitCodes.CATCH_ALL_GENERAL_ERROR_EXIT_CODE);
```
Comment (dizzzz) [inline src/org/exist/launcher/Launcher.java]: still commented out? probably a global search-replace ?

Comment (adamretter) [inline src/org/exist/launcher/Launcher.java]: Yes that should still be commented out :-)

File: src/org/exist/backup/ExportMain.java (hunk 20 above)
Comment (adamretter) [inline src/org/exist/backup/ExportMain.java]: That code has been fixed now I think

File: src/org/exist/start/start.config [hunk 22]
```diff
@@ -41,7 +41,7 @@ lib/core/commons-collections-%latest%.jar          always
 lib/core/commons-io-%latest%.jar                   always
 lib/core/commons-logging-%latest%.jar              always
 lib/core/commons-pool-%latest%.jar                 always
-lib/core/excalibur-cli-%latest%.jar                mode != jetty
+lib/core/jargo-%latest%.jar                        mode != jetty
```
Comment (dizzzz) [inline src/org/exist/start/start.config]: yuhuuu!

File: src/org/exist/client/InteractiveClient.java (hunk 18 above)
Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: @shabanovd Yes because the old code is redundant

File: src/org/exist/util/ArgumentUtil.java [hunk 23]
```diff
@@ -0,0 +1,83 @@
+package org.exist.util;
```
Comment (dizzzz) [inline src/org/exist/util/ArgumentUtil.java]: You might want to consider to contribute this back to the original project?

File: src/org/exist/backup/ExportMain.java (hunk 20 above)
Comment (dizzzz) [inline src/org/exist/backup/ExportMain.java]: yes, thnx

File: src/org/exist/client/InteractiveClient.java (hunk 19 above)
Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: This code was destined for eXist 3.0 but got left behind due to the "outrage about" Guava.

I think we can use some common sense here. As we know eXist does not have a proper public API, and if it did this certainly wouldn't be part of it. This method is called from `main` and it makes no sense to call it from elsewhere, it's a legacy mistake that it is not `private`.

File: src/org/exist/util/ArgumentUtil.java (hunk 23 above)
Comment (adamretter) [inline src/org/exist/util/ArgumentUtil.java]: Thanks, but I am not sure they are interested. eXist args are kinda a mess, as we don't use `commands` which is kinda what is expected these days and certainly what the Jargo design would typically expect.

**Comment by adamretter** [comment]
@dizzzz Thank you very much :-)

File: src/org/exist/client/InteractiveClient.java (hunk 19 above)
Comment (dizzzz) [inline src/org/exist/client/InteractiveClient.java]: well things moved from method to constructor. Since the parameters are from a different origin,  the API changes. Not sure if it is a problem for e.g. the webstart client?

Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: API change is API change! I did not say anything on `private` methods changes but all `public` or `protected` methods are public eXist API.

File: src/org/exist/client/InteractiveClient.java (hunk 18 above)
Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: nope, I think that you are not right here. Will your code produce same result for `windows`, `linux` and `osx` with `windows` style path?

**Comment by adamretter** [comment]
So what do you suggest instead that would make sense here?

File: src/org/exist/client/InteractiveClient.java (hunk 19 above)
Comment (dizzzz) [inline src/org/exist/client/InteractiveClient.java]: @shabanovd strictly you are right, but I don't mind here... 

So actually you say the full commandline handling should be done in the 'run(String...)' method? indeed that would not change the API.

Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: @dizzzz the problem that someone may build it own "tool" and depend on this method, we don't know for sure.

If we spend time to define rules we must follow it.

Comment (joewiz) [inline src/org/exist/client/InteractiveClient.java]: @shabanovd Sorry, this a little off topic from the issue at hand, but I'm just curious: 

The rules you mention, where can I find these?

You can imagine I'm seeing this as an opportunity for discussion, so I'd be grateful to consult whatever documentation there already is in this area.

Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]:
```
1. `MAJOR` versions contain incompatible API changes, including changes to the on-disk format of the database;
2. `MINOR` versions add functionality or deprecate API functions, without breaking backward compatibility; and
3. `PATCH` versions contain only backwards-compatible bug fixes.
```

Because of this change next eXist version must be 4.0.0 base on rule above: `versions contain incompatible API changes`.

File: src/org/exist/backup/ExportMain.java (hunk 20 above)
Comment (shabanovd) [inline src/org/exist/backup/ExportMain.java]: @adamretter I see that code stay same, where it was fixed?

File: src/org/exist/util/ArgumentUtil.java (hunk 23 above)
Comment (shabanovd) [inline src/org/exist/util/ArgumentUtil.java]: Missing header?

File: src/org/exist/util/ArgumentUtil.java (hunk 8 above)
Comment (shabanovd) [inline src/org/exist/util/ArgumentUtil.java]: better option to get default value from `argument`. Is it possible?

File: src/org/exist/client/InteractiveClient.java (hunk 19 above)
Comment (dizzzz) [inline src/org/exist/client/InteractiveClient.java]: Specific for this case, IMO, I don't see an issue....

Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: Then I have no idea why I spend time to discuss versions statement if we just going to violate it. My impression was that we going to define rules loud & clear to follow it (in all cases).

What is the problem to do not make this API change?
Why "big change" allow to violate versions statement?

I simply do not understand why this issue that easy to fix get so huge resistance.

File: src/org/exist/client/InteractiveClient.java (hunk 18 above)
Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: Yes, testing here shows it does

File: src/org/exist/client/InteractiveClient.java (hunk 19 above)
Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: Okay I changed the API back

File: src/org/exist/util/ArgumentUtil.java (hunk 23 above)
Comment (adamretter) [inline src/org/exist/util/ArgumentUtil.java]: I just added it

File: src/org/exist/client/InteractiveClient.java (hunk 19 above)
Comment (joewiz) [inline src/org/exist/client/InteractiveClient.java]: @shabanovd Ah, the versioning proposal was the source, got it.

**Comment by adamretter** [comment]
@shabanovd Can we now please finally merge this?

File: src/org/exist/client/InteractiveClient.java (hunk 18 above)
Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: Nope, here simple test for you:
```
  public static void main(String[] args) {
    test("test/test.file");
    test("test\\test.file");
  }

  static void test(String name) {
    o(name);
    n(name);
  }

  static void o(String name) {
    String fileName = name.replace('/', java.io.File.separatorChar).replace('\\', java.io.File.separatorChar);
    System.out.println(name + " > old > " + Paths.get(fileName).normalize());
  }

  static void n(String name) {
    System.out.println(name + " > new > " + Paths.get(name).normalize());

    System.out.println("file name = " + Paths.get(name).normalize().getFileName());
  }
```
and result on my mac:
```
test/test.file > old > test/test.file
test/test.file > new > test/test.file
file name = test.file
test\test.file > old > test/test.file
test\test.file > new > test\test.file
file name = test\test.file
```

Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: I think you misunderstand. `java.nio.file.Paths` and `java.nio.file.Path` are platform aware. So if you run your own test code (from above) on Windows, you will get different output (which is correct for Windows):

```
test/test.file > old > test\test.file
test/test.file > new > test\test.file
file name = test.file
test\test.file > old > test\test.file
test\test.file > new > test\test.file
file name = test.file
```

Which is why I told you that testing my code (which I did on both Mac and Windows) produces the correct results.

Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: `java.nio.file.Paths and java.nio.file.Path are platform aware` - that exactly what I point to you!

for windows result:
```
test/test.file > old > test\test.file
test/test.file > new > test/test.file
file name = test/test.file
test\test.file > old > test\test.file
test\test.file > new > test\test.file
file name = test.file
```

You didn't run test on windows, right?

My point that before scripts did work with "windows paths" on linux & "linux  paths" on windows, but that PR change this behavior.

I don't know why it was coded that way, just point to that fact.

Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: Yes I ran the test on Windows 10 with JDK 7 and 8 and got the results I pasted above

Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: Here is a screenshot to prove it for you - http://static.adamretter.org.uk/virtual-box-windows10.png

Comment (dizzzz) [inline src/org/exist/client/InteractiveClient.java]: @shabanovd how can your result be `file name = test/test.file` in stead of `file name = test.file` ?

Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: @adamretter ok, so things ever worse:

my windows (jdk1.8.0_25):
```
test/test.file > old > test\test.file
test/test.file > new > test\test.file
file name = test.file
test\test.file > old > test\test.file
test\test.file > new > test\test.file
file name = test.file
```

my osx (jdk1.8.0_77):
```
test/test.file > old > test/test.file
test/test.file > new > test/test.file
file name = test.file
test\test.file > old > test/test.file
test\test.file > new > test\test.file
file name = test\test.file
```

my linux (1.8.0_101):
```
test/test.file > old > test/test.file
test/test.file > new > test/test.file
file name = test.file
test\test.file > old > test/test.file
test\test.file > new > test\test.file
file name = test\test.file
```

Comment (shabanovd) [inline src/org/exist/client/InteractiveClient.java]: @wolfgangmm we need you because `fileName.replace('/', java.io.File.separatorChar).replace('\\', java.io.File.separatorChar);` here from [1st commit](https://github.com/adamretter/exist/commit/7ac94759168fc331419ad4d5777c5c07187d208e#diff-e50a0f4a572caabd4804d0cd924aa666R1088).

Will you able to recall why it was done this way? (with hope)

Comment (adamretter) [inline src/org/exist/client/InteractiveClient.java]: @shabanovd I have been trying to tell you that this isn't a problem. Only Windows needs to handle Windows file paths!
